scalalib/src/mill/javalib/android/AndroidAppBundle.scala

```diff
@@ -0,0 +1,214 @@
+   */
+  def bundleToolUrl: T[String] = Task {
+    "https://github.com/google/bundletool/releases/download/1.17.2/bundletool-all-1.17.2.jar"
+  }
+
+  /**
+   * Compiles Android resources using AAPT2, linking them into a single archive
+   * to be included in the app bundle.
+   *
+   * @return PathRef to the compiled resources directory.
+   */
+  def androidResources: T[PathRef] = Task {
+    // Define paths for the AAPT2 jar and extracted tool
+    val aapt2Jar = Task.dest / "aapt2.jar"
+    val aapt2 = Task.dest / "aapt2"
+    val compiledResDir = Task.dest / "compiled"
+
+    // Get resource directories and filter for existing ones
+    val resourceDirs = resources().map(_.path).filter(os.exists)
+
+    // Download and extract AAPT2 tool
+    os.write(aapt2Jar, requests.get(aapt2Url()).bytes)
+    os.unzip(aapt2Jar, Task.dest)
+    os.call(("chmod", "+x", aapt2)) // Make AAPT2 executable
+    os.makeDir.all(compiledResDir) // Ensure compiled resource directory exists
+
+    // Compile each resource directory using AAPT2
+    var count = 0
+    val compiledZips = resourceDirs.map { resDir =>
+      val outputZip = compiledResDir / s"${resDir.last}-${count}.zip"
+      count += 1
+      os.call((
+        aapt2,
+        "compile",
+        "--dir",
+        resDir,
+        "-o",
+        outputZip
+      ))
+      outputZip
+    }
+
+    // Separate main resource zip and library zips for later linking
+    val ResourceZip = compiledZips.find(_.toString.contains("resources"))
+    val libzips = compiledZips.filterNot(_.toString.contains("resources"))
+    val compiledlibs = libzips.flatMap(zip => Seq("-R", zip.toString))
+
+    // Link resources and generate res.apk
+    os.call(
+      Seq(
+        aapt2.toString, // AAPT2 executable path
+        "link",
+        "-I",
+        androidSdkModule().androidJarPath().path.toString, // Android SDK jar
+        "--auto-add-overlay",
+        "--proto-format" // Format and overlay options
+      ) ++ compiledlibs ++ Seq(
+        "--manifest",
+        androidManifest().path.toString, // Use provided manifest
+        "--java",
+        Task.dest.toString, // Output generated R.java
+        "-o",
+        s"${Task.dest / "res.apk"}", // Output APK path
+        ResourceZip.map(_.toString).getOrElse("")
+      )
+    )
+
+    PathRef(Task.dest) // Return compiled resource directory as PathRef
+  }
+
+  /**
+   * Packages compiled resources and dex files into a zip for app bundle generation.
+   *
+   * @return PathRef to the created bundle zip.
+   */
+  def androidBundleZip: T[PathRef] = Task {
+    val dexFile = androidDex().path
+    val baseDir = Task.dest / "base"
+
+    // Create directory structure for Android bundle
+    Seq("manifest", "dex").foreach(d => os.makeDir.all(baseDir / d))
+
+    // Copy dex and resource files into the bundle directory
+    os.copy(dexFile, Task.dest / dexFile.last)
+    os.unzip(Task.dest / os.up / "androidResources.dest" / "res.apk", Task.dest)
```

Comment:
We should never use `Task.dest / os.up`, as a task should only write to its own `dest` folder. Not sure what you are trying to do here, but if you want to write to `androidResources.dest` then the `def androidResources` task should do the writing

Reply:
Actually i want the reas.apk which was generated during the android resources process and in need to add dex file into it and as we know before running androiddex we will always run the android resources so think there will be no issue with the usage of os.up 

But still if you say i will call the android resources task rather than using hard-coded path to res.apk

Reply:
In that case you should copy the `res.apk` in this task's `Test.dest` folder and modify your own copy